File: PR/lab5/selectSort.c
#include <stdio.h>
#include <stdlib.h>

int selectionComparatorUp(int first, int second) {
    return first < second;
}
int selectionComparatorDown(int first, int second) {
    return first > second;
}
int operation(int (*op)(int, int), int a, int b) {
    return op(a,b);
}
void selectionSort(int array[], int size, FILE *outputFile) {
    for (int i = 0; i < size - 1; i++) {
        int minIndex = i;
        for (int j = i + 1; j < size; j++) {
            if (operation(selectionComparatorUp, array[j], array[minIndex]) == 1) {
                minIndex = j;
            }
        }
        printf("Sorted %d/%d\n", i, size);
        if (minIndex != i) {
            int temp = array[minIndex];
            array[minIndex] = array[i];
            array[i] = temp;
        }
    }
    printf("Sorted");
    for (int i = 0; i < size; i++) {
        fprintf(outputFile, "%d ", array[i]);
    }
} 
int main(void) {
    long long n;
    FILE *outputFile = fopen("output.txt", "w");
    FILE *file = fopen("test.txt", "r");
    fscanf(file, "%lld", &n);
    int* array = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        fscanf(file, "%d", &array[i]);
    }
    selectionSort(array, n, outputFile);
    fclose(file);
    fclose(outputFile);
    return 0;
}

File: PR/lab5/selectSort.pas
program lab1;

type
  comparator = function(first, second:longint): Boolean;

function selectionComparatorUp(first, second: longint): Boolean;
begin
  selectionComparatorUp := first < second;
end;

function selectionComparatorDown(first, second: longint): Boolean;
begin
  selectionComparatorDown := first > second;
end;

procedure selectionSort(arr: array of longint; size: longint; var outputFile: TextFile; comp:comparator);
var
  i, j, minIndex, temp: longint;
begin
  for i := 0 to size - 2 do
  begin
    minIndex := i;
    for j := i + 1 to size - 1 do
    begin
      if comp(arr[j], arr[minIndex]) then
      begin
        minIndex := j;
      end;
    end;
    if minIndex <> i then
    begin
      temp := arr[minIndex];
      arr[minIndex] := arr[i];
      arr[i] := temp;
    end;
  end;
  for i := 0 to size - 1 do
  begin
    Write(outputFile, arr[i]);
    if i < size-1 then Write(outputFile, ' ');
  end;
  Writeln(outputFile, '');
end;

var
  n: longint;
  outputFile: TextFile;
  inputFile: TextFile;
  arr: array of longint;
  i: longint;
begin
  Assign(outputFile, 'output.txt');
  Rewrite(outputFile);
  Assign(inputFile, 'input.txt');
  Reset(inputFile);
  Read(inputFile, n);
  SetLength(arr, n);
  for i := 0 to n - 1 do
  begin
    Read(inputFile, arr[i]);
  end;
  selectionSort(arr, n, outputFile, @selectionComparatorUp);
  Close(inputFile);
  Close(outputFile);
end.
